### PR TITLE
enhancement(remap): add merge and merge equals operations

### PR DESCRIPTION
**Pull request by FungusHumungus**
This adds the merge (`|`) and merge equals (`|=`) operators.

The `|=` operator works by essentially rewriting the syntax of `a |= b` into `a = a | b` and `a, err |= b` into `a, err = a | b`.

It opens up the door to other operators such as `+=` should we wish to incorporate that.

It is unfortunate that this has resulted in the entire AST needing to be Clonable, but I will defer that to a future issue.

Signed-off-by: Stephen Wakely <fungus.humungus@gmail.com>

### CODE REVIEW

File: lib/vrl/compiler/src/compiler.rs [hunk 1]
```diff
@@ -206,21 +206,66 @@ impl<'a> Compiler<'a> {
...
+                                    Node::new(span, ast::Opcode::Union),
+                                    expr,
+                                ),
+                            ))),
+                        ));
+
+                        let node = Variant::Infallible { ok, err, expr };
+                        Node::new(span, node)
+                    }
+                }
```
Comment (FungusHumungus) [inline lib/vrl/compiler/src/compiler.rs:266]: This is where the syntax rewriting is happening. I'm sure theres a good chunk of code duplication that can be refactored. I wonder if it would be useful to enable more library support for syntax rewriting, depending on if it would be useful elsewhere too. Perhaps, ultimately we could have a stage before the compiler that does this.

This could also be extended for `+=` etc..

File: src/mapping/query/arithmetic.rs [hunk 2]
```diff
@@ -21,6 +21,7 @@ pub(in crate::mapping) enum Operator {
     LessOrEqual,
     And,
     Or,
+    Union,
```
Comment (FungusHumungus) [inline src/mapping/query/arithmetic.rs:24]: I should rename this `Merge`.

File: lib/vrl/parser/src/lex.rs [hunk 3]
```diff
@@ -260,6 +260,7 @@ pub enum Token<S> {
     Escape,
 
     Equals,
+    MergeEquals,
     Bang,
     Pipe,
```
Comment (FungusHumungus) [inline lib/vrl/parser/src/lex.rs:265]: The `Pipe` and merge `|` are the same symbol. This needs tidying up. Both path coalescing and merging seem to parse correctly, but there is the potential for conflicts and confusion..

File: lib/vrl/compiler/src/compiler.rs [hunk 4]
```diff
@@ -206,21 +206,66 @@ impl<'a> Compiler<'a> {
         use ast::Assignment::*;
 
         self.state.snapshot();
+        let (_span, assignment) = node.take();
```
Comment (JeanMertz) [inline lib/vrl/compiler/src/compiler.rs:209]:
```suggestion
        let assignment = node.into_inner();
```

File: lib/vrl/compiler/src/compiler.rs (hunk 1 above)
Comment (JeanMertz) [reply lib/vrl/compiler/src/compiler.rs:266]: I concur, we need an [intermediate representation](https://en.wikipedia.org/wiki/Intermediate_representation) layer 😛

File: lib/vrl/parser/src/ast.rs [hunk 5]
```diff
@@ -742,13 +747,36 @@ pub enum Assignment {
...
+        match self {
+            Assign => write!(f, "="),
+            Merge => write!(f, "|="),
+        }
+    }
+}
+
+impl fmt::Debug for AssignmentOp {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Display::fmt(self, f)
```
Comment (JeanMertz) [inline lib/vrl/parser/src/ast.rs:769]: Most (all?) debug impls in the AST do something like `AssignmentOp(=)` and `AssigmentOp(|=)`.

We didn't for the assignment operator before because you could only use `=`, but now that's changed, it'd be nice to stick to that convention.

File: lib/vrl/parser/src/ast.rs [hunk 6]
```diff
@@ -758,18 +786,38 @@ impl fmt::Debug for Assignment {
...
     Noop,
     Query(Query),
     Internal(Ident, Option<Path>),
-    External(Option<Path>),
+}
+
+impl AssignmentTarget {
+    pub fn into_expr(self, span: &Span) -> Expr {
+        match self {
+            AssignmentTarget::Noop => Expr::Literal(Node::new(span.clone(), Literal::Null)),
```
Comment (JeanMertz) [inline lib/vrl/parser/src/ast.rs:807]: Isn't `Span: Copy`? I wouldn't have expected the need to clone them.

File: lib/vrl/parser/src/ast.rs [hunk 7]
```diff
@@ -758,18 +786,38 @@ impl fmt::Debug for Assignment {
...
+#[derive(Clone, PartialEq)]
 pub enum AssignmentTarget {
     Noop,
     Query(Query),
     Internal(Ident, Option<Path>),
-    External(Option<Path>),
+}
+
+impl AssignmentTarget {
+    pub fn into_expr(self, span: &Span) -> Expr {
```
Comment (JeanMertz) [inline lib/vrl/parser/src/ast.rs:805]: Since `Span: Copy`, you don't actually need to pass a reference here.

File: lib/vrl/parser/src/ast.rs [hunk 8]
```diff
@@ -758,18 +786,38 @@ impl fmt::Debug for Assignment {
...
+            AssignmentTarget::Internal(ident, Some(path)) => Expr::Query(Node::new(
+                span.clone(),
+                Query {
+                    target: Node::new(span.clone(), QueryTarget::Internal(ident)),
+                    path: Node::new(span.clone(), path),
+                },
+            )),
+            AssignmentTarget::Internal(ident, None) => {
+                Expr::Variable(Node::new(span.clone(), ident))
+            }
```
Comment (JeanMertz) [inline lib/vrl/parser/src/ast.rs:818]: As discussed on Discord. I prefer we keep using `Internal` and `External` for now, even though some recent parser changes have made them obsolete. We should remove them, but I'd rather do that in a follow-up PR that looks at the bigger picture and looks into simplifying the parser as well, now that these types are no longer needed.

(we can create an issue to track this work)

File: lib/vrl/parser/src/ast.rs [hunk 9]
```diff
@@ -806,7 +850,7 @@ impl fmt::Debug for AssignmentTarget {
 // query
 // -----------------------------------------------------------------------------
 
-#[derive(PartialEq)]
+#[derive(Clone, PartialEq)]
```
Comment (JeanMertz) [inline lib/vrl/parser/src/ast.rs:853]: The added `Clone`s is unfortunate.

As discussed on Discord with @FungusHumungus, it should be possible to remove these, but we'll defer that until after 0.12 is released. Could you create a tracking issue for this @FungusHumungus?

File: lib/vrl/parser/src/parser.lalrpop [hunk 10]
```diff
@@ -294,6 +301,7 @@ Compare: Expr = {
 };
 
 Add: Expr = {
+    Op<Add, "|", Factor>,
```
Comment (JeanMertz) [inline lib/vrl/parser/src/parser.lalrpop:304]: Might want to move this into a separate rule named `Merge` just for consistency?

File: lib/vrl/tests/tests/expressions/arithmetic/union/multiple_maps.vrl [hunk 11]
```diff
@@ -0,0 +1,3 @@
+#result: { "foo": 1, "bar": 2, "zub": 3, "zoob": 4 }
+
+{ "foo": 1, "bar": 99 } | { "bar": 2 } | { "zub": 3, "zoob": 99 } | { "zoob": 4 }
```
Comment (JeanMertz) [inline lib/vrl/tests/tests/expressions/arithmetic/union/multiple_maps.vrl:3]: Would be good to add a test to see what happens with deep merges.

File: lib/vrl/parser/src/parser.lalrpop [hunk 12]
```diff
@@ -40,6 +40,7 @@ extern {
         "?" => Token::Question,
         "|" => Token::Pipe,
```
Comment (JeanMertz) [inline lib/vrl/parser/src/parser.lalrpop:41]: I think we can remove this (unused?) token, right?

File: lib/vrl/compiler/src/expression/op.rs [hunk 13]
```diff
@@ -100,6 +101,12 @@ impl Expression for Op {
             // ... || ...
             Or => merged_def,
 
+            // {} | {}
+            Union if lhs_kind.is_object() && rhs_kind.is_object() => merged_def,
+
+            // ... | ...
+            Union  => merged_def.fallible(),
```
Comment (JeanMertz) [inline lib/vrl/compiler/src/expression/op.rs:108]: With https://github.com/timberio/vector/pull/6471 merged, it would be nice to add a specific compile-time error for this case, that explains what's going on here.

Right now, it would just result in a generic "this expression can fail, resolve the failure to guarantee runtime success".

File: lib/vrl/parser/src/ast.rs (hunk 6 above)
Comment (FungusHumungus) [reply lib/vrl/parser/src/ast.rs:807]: Oh, nice!

File: lib/vrl/tests/tests/expressions/arithmetic/union/multiple_maps.vrl (hunk 11 above)
Comment (FungusHumungus) [reply lib/vrl/tests/tests/expressions/arithmetic/union/multiple_maps.vrl:3]: These merges are only ever shallow. If you want deep, you need the `merge` function. Should it be the other way round?

File: lib/vrl/parser/src/parser.lalrpop (hunk 12 above)
Comment (FungusHumungus) [reply lib/vrl/parser/src/parser.lalrpop:41]: I recall trying, but think it is still used by the coalescing in paths.. I'll try again. It needs cleaning up regardless..

File: lib/vrl/tests/tests/expressions/arithmetic/union/multiple_maps.vrl (hunk 11 above)
Comment (JeanMertz) [reply lib/vrl/tests/tests/expressions/arithmetic/union/multiple_maps.vrl:3]: No, I meant we should add a test that validates that the merges are indeed shallow 😄

File: lib/vrl/compiler/src/expression/op.rs [hunk 14]
```diff
@@ -263,6 +290,23 @@ impl DiagnosticError for Error {
                 Label::context("this expression never resolves", rhs_span),
                 Label::context("remove this error coalesce operation", op_span),
             ],
+            MergeNonObjects { lhs_span, rhs_span } => {
+                let mut labels = Vec::new();
+                if let Some(lhs_span) = lhs_span {
+                    labels.push(Label::context(
+                        "this expression must resolve to an object",
+                        lhs_span,
+                    ));
```
Comment (JeanMertz) [inline lib/vrl/compiler/src/expression/op.rs:299]: I propose changing this to `Label::primary`.

Basically, every error should (almost) always have at least _one_ primary span, which is coloured red, and points to the issue.

Then, if need be, one or more context spans can be used to explain what's going on and how the user can solve the issue.

File: lib/vrl/compiler/src/expression/op.rs [hunk 15]
```diff
@@ -263,6 +290,23 @@ impl DiagnosticError for Error {
...
+                }
+                if let Some(rhs_span) = rhs_span {
+                    labels.push(Label::context(
```
Comment (JeanMertz) [inline lib/vrl/compiler/src/expression/op.rs:302]: Same here, re: making this `Label::primary`

File: lib/vrl/compiler/src/value/error.rs [hunk 16]
```diff
@@ -55,6 +55,9 @@ pub enum Error {
 
     #[error("unable to compare {0} <= {1}")]
     Le(Kind, Kind),
+
+    #[error("unable to merge {0} <= {1}")]
```
Comment (JeanMertz) [inline lib/vrl/compiler/src/value/error.rs:59]:
```suggestion
    #[error("unable to merge {1} into {0}")]
```

File: lib/vrl/parser/src/parser.lalrpop [hunk 17]
```diff
@@ -244,13 +246,18 @@ Assignment: Node<Assignment> = {
 
 AssignmentSingle: Assignment = {
     <target: Sp<AssignmentTarget>> "="
-        <expr: Box<Expr>> => Assignment::Single { target, expr },
+        <expr: Box<Expr>> => Assignment::Single { target, op: AssignmentOp::Assign, expr },
+    <target: Sp<AssignmentTarget>> "|="
+        <expr: Box<Expr>> => Assignment::Single { target, op: AssignmentOp::Merge, expr },
```
Comment (JeanMertz) [inline lib/vrl/parser/src/parser.lalrpop:251]: No need to do this here, but I think we should be able to merge these two and have a new `AssignmentOp` non-terminal that matches either on `=` and `|=` and then returns the correct enum variant.

Similar to how we have an `AssignmentTarget` non-terminal as well.

(same applies to the infallible assignment below)

File: lib/vrl/tests/tests/expressions/arithmetic/merge/invalid.vrl [hunk 18]
```diff
@@ -0,0 +1,29 @@
+# result:
```
Comment (JeanMertz) [inline lib/vrl/tests/tests/expressions/arithmetic/merge/invalid.vrl:1]: Pro-tip: if you look at the other `invalid.vrl` examples, you can do the following:

```ruby
# result: ~
#
# can only merge objects
# can only merge objects
# can only merge objects
```

(note the `~`)

Basically, this allows you to validate that an error with that title occurred, but prevents you from having to update the UI tests every time a minor detail in the diagnostic is updated.

We still have `tests/diagnostics` where you'd add _one_ full example of this error to make sure any UI changes are still caught there.

File: lib/vrl/tests/tests/expressions/arithmetic/merge/type_defs_wrong.vrl [hunk 19]
```diff
@@ -0,0 +1,33 @@
+# result:
```
Comment (JeanMertz) [inline lib/vrl/tests/tests/expressions/arithmetic/merge/type_defs_wrong.vrl:1]: Same here, re: using `~`

File: lib/vrl/tests/tests/expressions/assignment/merge_assignment_wrong_types.vrl [hunk 20]
```diff
@@ -0,0 +1,53 @@
+# result:
```
Comment (JeanMertz) [inline lib/vrl/tests/tests/expressions/assignment/merge_assignment_wrong_types.vrl:1]: Same here: re `~`

**Review by JeanMertz — APPROVED** [review]
Sweet!

Left a few comments/nits, but otherwise this look good to go 👌

File: lib/vrl/tests/tests/expressions/arithmetic/merge/invalid.vrl (hunk 18 above)
Comment (FungusHumungus) [reply lib/vrl/tests/tests/expressions/arithmetic/merge/invalid.vrl:1]: Nice.

File: lib/vrl/parser/src/parser.lalrpop [hunk 21]
```diff
@@ -242,15 +244,22 @@ Assignment: Node<Assignment> = {
...
+AssignmentOp: AssignmentOp = {
+    "=" => AssignmentOp::Assign,
+    "|=" => AssignmentOp::Merge,
+}
+
 AssignmentSingle: Assignment = {
-    <target: Sp<AssignmentTarget>> "="
-        <expr: Box<Expr>> => Assignment::Single { target, expr },
+    <target: Sp<AssignmentTarget>>
+        <op: AssignmentOp>
```
Comment (JeanMertz) [inline lib/vrl/parser/src/parser.lalrpop:254]: One more note (again, nothing you need to fix in this PR):

In general, I've tried to get as much span information from the parser as possible, so that our AST consistently provides span details for anything that's relevant for consumers of the AST.

In this case, the `=` token wasn't relevant before because there was no other alternative to use. Now that it can be either `=` or `|=`, it makes sense to include span information for the assignment operator as well.

We can tackle this in a follow-up PR after 0.12 ships though, as we don't actually need that span right now.

File: lib/vrl/stdlib/src/parse_key_value.rs [hunk 22]
```diff
@@ -98,7 +98,7 @@ impl Expression for ParseKeyValueFn {
     }
 
     fn type_def(&self, _: &state::Compiler) -> TypeDef {
-        TypeDef::new().infallible().object::<(), Kind>(map! {
+        TypeDef::new().fallible().object::<(), Kind>(map! {
```
Comment (pablosichert) [inline lib/vrl/stdlib/src/parse_key_value.rs:101]: This change seems unrelated and also might not be correct, as it can fail at runtime e.g. when the input ended unexpectedly?

File: lib/vrl/compiler/src/compiler.rs [hunk 23]
```diff
@@ -204,26 +204,69 @@ impl<'a> Compiler<'a> {
         })
     }
 
+    /// Rewrites the ast for `a |= b` to be `a = a | b`.
+    fn rewrite_to_merge(
+        &mut self,
+        span: diagnostic::Span,
+        target: &Node<ast::AssignmentTarget>,
+        expr: Box<Node<ast::Expr>>,
+    ) -> Box<Node<Expr>> {
```
Comment (pablosichert) [inline lib/vrl/compiler/src/compiler.rs:213]: Nit: I'd let the caller box this one

**Review by pablosichert — APPROVED** [review]
Nice work!

File: lib/vrl/stdlib/src/parse_key_value.rs (hunk 22 above)
Comment (FungusHumungus) [reply lib/vrl/stdlib/src/parse_key_value.rs:101]: Yeah, `parse_key_value("ook")` will fail since there is no key=value. I needed a good function that could fail that returned a map to test the merge code, and noticed that the type def here was incorrectly specified, so I just fixed it to use in the tests..

File: lib/vrl/compiler/src/compiler.rs (hunk 23 above)
Comment (FungusHumungus) [reply lib/vrl/compiler/src/compiler.rs:213]: For what reason?

Comment (pablosichert) [reply lib/vrl/compiler/src/compiler.rs:213]: Simply because this function shouldn't be concerned about where the result is allocated. Hypothetically, if you wanted to allocate the result on the stack, this would prevent you from doing it.

This is just an argument about principle, not about this practical case. Hence why only a nit and up to your discretion to change it or not 😄

Comment (FungusHumungus) [reply lib/vrl/compiler/src/compiler.rs:213]: Ah yeah, that's a good point. I won't make that change here, just because it is only called from two places, and both places need to box it, but will bear that in mind for the future! Thanks.

File: lib/vrl/stdlib/src/parse_key_value.rs (hunk 22 above)
Comment (pablosichert) [reply lib/vrl/stdlib/src/parse_key_value.rs:101]: Oh, I somehow read the diff the other way around. All good 😄
